README
=Google Calendar Plugin For Redmine

Dependency:  TZInfo GEM  (http://tzinfo.rubyforge.org/)

==Description

A plugin to allow users to create a new Redmine tab and embed a Google Calendar using Google Embeddable Calendar iframe.

For help on how to obtain the calendar-information see http://www.google.com/calendar/embed/EmbedHelper_en.html.

==Installation

To install from git: 

(Git Version < 1.5.3)
  $ cd {RAILS_ROOT}/vendor/plugins 
  $ git clone git://github.com/mully/redmine_google_calendar.git

(Git Version >= 1.5.3 -- Because the "git-rails" gem uses Git Submodules introduced in ver. 1.5.3)
  $ sudo gem install git-rails
  $ git-rails install git://github.com/mully/redmine_google_calendar.git

==Plugin setup

1. Install the tzinfo gem if it does not already exist on your server:
   $ sudo gem install tzinfo
2. Install the plugin as described at http://www.redmine.org/wiki/redmine/Plugins (this plugin doesn't require migration).
3. Go to "Administration -> Custom fields" and create a Project custom field of the "long text" type, named 'Google Calendar IFrame' for example.
4. Go to "Administration -> Plugins -> Google Calendar Plugin > Configure" to configure the Google Calendar plugin:
   * select the project custom field which contains the Google Calendar-iframe; in this example select the newly created project custom field "Google Calendar IFrame"

==Enabling Calendar at project level

===Retrieve the Google Calendar embedding-code (<iframe>)

Go to the Google Calendar you want to embed:

1. Click on "Settings" in the upper left hand corner
2. Click on the "Calendars" tab
3. Select the calendar you want to use
4. In the "Embed This Calendar" section, copy the entire <iframe> code from the text box

(If this is a private calendar, you will have to replace the "src" in the <iframe> with the HTML found by clicking the blue "HTML" block in the "Private Address" section on the same page.)

===Add the Google Calendar embedding-code (<iframe>) in the Redmine project

In your project in Redmine, go to "Project settings" :

1. On the "Modules" tab: enable the module by checking "Google Calendar"
2. On the "Information" tab: paste the <iframe> code, that you have copied to your clipboard from Google, into the 'Google Calendar IFrame' project custom field you created in step 1

After that, a new item named "Calendar" should show up in the project menu. Click on this item to see your Google Calendar.


CHANGELOG
0.1.1 (14-06-2009):
    * Defect: Fix compatibility with Redmine trunk while keeping it compatible with Redmine 0.8.x
    * Defect: Fix development-mode compatibility with Redmine trunk while keeping it compatible with Redmine 0.8.x in development-mode
    * Defect: Currently selected project custom field for the iframe-source is not reflected in the view
    * Enhancement: Remove obsolete "net/imap"-requirement
    * Enhancement: Carry-out the plugin name consequently through all code(-comments)
    * Enhancement: Rename the "settings"-partial to prevent collision with other plugins
    * Enhancement: Structurize the code-layout (indention, spaces etc.) and remove obsolete code-comments
    * Enhancement: Update and improve the delivered documentation
    * Enhancement: More verbose logging while starting the plugin
    * Enhancement: Update and structurize the licensing documentation
    * Enhancement: Add license-file
    * Enhancement: Remove obsolete files and directories
    * Enhancement: Add changelog-file

0.1.0 (26-03-2008):
    * Feature: Added project-menu item (tab) highlighting

0.1.0 (12-03-2008):
    - Initial public release


app/controllers/google_calendar_controller.rb
#       iframe_tab_controller.rb
#       
#       Copyright 2008 Jim Mulholland <jim@squeejee.com>
#       
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#       
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#       
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, write to the Free Software
#       Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#       MA 02110-1301, USA.

require 'tzinfo'

class GoogleCalendarController < ApplicationController
  unloadable
  layout 'base'
  
  before_filter :find_project, :authorize
  
  def show
    @iframe_text = IframeText.get_iframe_text(@project)
    
    unless User.current.time_zone.nil?
        time_zone = tzinfo_from_offset(User.current.time_zone.utc_offset)
        
        #If "pvttk" string is not in the iframe, this is a public calendar
        if @iframe_text[/pvttk/].nil?
            #Substitute in the current timezone for public calendar
            @iframe_text.sub!(/ctz=\S*"/, "ctz=#{time_zone.name}\"")
        else
            #Substitute in the current timezone for private calendar
            @iframe_text.sub!(/ctz=\S*&/, "ctz=#{time_zone.name}&")
        end
    end
  end
  

private
  def find_project   
    @project = Project.find(params[:id])
  rescue ActiveRecord::RecordNotFound
    render_404
  end
  
  def tzinfo_from_offset(offset_in_seconds)
    
    #Search For US Timezones First
    us = TZInfo::Country.get('US')
    
    us.zone_info.each do |tz|
        if tz.timezone.current_period.utc_offset.to_i == offset_in_seconds        
            return tz.timezone
        end
    end
    
    TZInfo::Timezone.all.each do |tz2|
        if tz2.current_period.utc_offset.to_i == offset_in_seconds
            return tz2
        end
    end
    return nil
   end
end



init.rb
# Redmine Google Calendar plugin
# 
# Copyright 2009 Mischa The Evil <mischa_the_evil@hotmail.com>
# Copyright 2008 Jim Mulholland <jim@squeejee.com>
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.

require 'redmine'

RAILS_DEFAULT_LOGGER.info 'Starting Google Calendar plugin 0.1.1'

# Redmine Google Calendar plugin
Redmine::Plugin.register :redmine_google_calendar do
  name 'Google Calendar Plugin'
  author 'Jim Mulholland'
  description 'A plugin to allow users to add a new tab with an embedded Google Calendar Iframe.'
  version '0.1.1'
  
  # This plugin contains settings
  settings :default => {
    'iframe_text' => ''
  }, :partial => 'settings/googlecalendar_settings'
  
  # This plugin adds a project module
  # It can be enabled/disabled at project level (Project settings -> Modules)
  project_module :google_calendar do
    # This permission has to be explicitly given
    # It will be listed on the permissions screen
    permission :view_google_calendar_tab, {:google_calendar => :show}
  end
  
  # A new item is added to the project menu
  menu :project_menu, :google_calendar, :controller => 'google_calendar', :action => 'show'
end
